contracts/zerodev/account/kernel/WalletKernel.sol

```diff
@@ -0,0 +1,189 @@
+        if (!success) {
+            assembly {
+                revert(add(ret, 32), mload(ret))
+            }
+        }
+    }
+
+
+    function validateUserOp(UserOperation calldata userOp, bytes32 userOpHash, uint256 missingAccountFunds)
+    external returns (uint256 validationData) {
+        require(UserOperationLib.checkUserOpOffset(userOp), "userOp: invalid offset");
+        require(msg.sender == entryPoint, "account: not from entryPoint");
+        if(userOp.signature.length == 65){
+            validationData = _validateUserOp(userOp, userOpHash);
+        } else if(userOp.signature.length > 97) {
+            // userOp.signature = address(plugin) + validUntil + validAfter + pluginData + pluginSignature
+            address plugin = address(bytes20(userOp.signature[0:20]));
+            uint48 validUntil = uint48(bytes6(userOp.signature[20:26]));
+            uint48 validAfter = uint48(bytes6(userOp.signature[26:32]));
+            bytes memory signature = userOp.signature[32:97];
+            (bytes memory data, ) = abi.decode(userOp.signature[97:], (bytes, bytes));
+
+            bytes32 digest = _hashTypedDataV4(keccak256(abi.encode(
+                keccak256("ValidateUserOpPlugin(address sender,uint48 validUntil,uint48 validAfter,address plugin,bytes data)"), // we are going to trust plugin for verification
+                userOp.sender,
+                validUntil,
+                validAfter,
+                plugin,
+                keccak256(data)
+            )));
+
+            address signer = ECDSA.recover(digest, signature);
+            if(getWalletKernelStorage().owner != signer) {
+                return SIG_VALIDATION_FAILED;
+            }
+            bytes memory ret = _delegateToPlugin(
+                plugin,
+                userOp,
+                userOpHash,
+                missingAccountFunds
+            );
+            bool res = abi.decode(ret, (bool));
+            if(res) {
+                return SIG_VALIDATION_FAILED;
+            }
+            validationData = _packValidationData(!res, validUntil, validAfter);
+        } else {
+            return SIG_VALIDATION_FAILED;
+        }
+        if(missingAccountFunds > 0) { // we are going to assume signature is valid at this point
+            (bool success, ) = msg.sender.call{value: missingAccountFunds}("");
+            (success);
+            return validationData;
+        }
+    }
+
+    function _validateUserOp(UserOperation calldata userOp, bytes32 userOpHash)
+    internal returns (uint256 validationData) {
+        bytes32 hash = ECDSA.toEthSignedMessageHash(userOpHash);
+        address recovered = ECDSA.recover(hash,userOp.signature);
+        WalletKernelStorage storage ws = getWalletKernelStorage();
+        if (ws.owner != recovered) {
+            return SIG_VALIDATION_FAILED;
+        }
+
+        if (userOp.initCode.length == 0) {
+            if(ws.nonce++ != userOp.nonce) {
+                return SIG_VALIDATION_FAILED;
```

Comment:
Should we be using `SIG_VALIDATION_FAILED` for nonce errors?

Reply:
wdym? we are using it now right?

Reply:
Yeah I meant like, maybe we shouldn't?  Cuz I thought `SIG_VALIDATION_FAILED` is only for signature errors

Reply:
makes sense, i thought that it should not revert on any case but seems it's only for signatures.
let's revert on this one